web/src/engine/osk/src/keyboard-layout/oskLayer.ts

```diff
@@ -132,42 +136,63 @@ export default class OSKLayer {
       this.spaceBarKey.spec.text = displayName;
 
-      // It sounds redundant, but this dramatically cuts down on browser DOM processing;
-      // but sometimes innerText is reported empty when it actually isn't, so set it
-      // anyway in that case (Safari, iOS 14.4)
-      if (spacebarLabel.innerText != displayName || displayName == '') {
-        spacebarLabel.innerText = displayName;
+      return () => {
+        // It sounds redundant, but this dramatically cuts down on browser DOM processing;
+        // but sometimes innerText is reported empty when it actually isn't, so set it
+        // anyway in that case (Safari, iOS 14.4)
+        if (spacebarLabel.innerText != displayName || displayName == '') {
+          spacebarLabel.innerText = displayName;
+        }
       }
     }
     catch (ex) { }
   }
 
-  public refreshLayout(vkbd: VisualKeyboard, layoutParams: LayerLayoutParams) {
-    const layerHeight = layoutParams.keyboardHeight;
-    this.showLanguage(layoutParams.spacebarText);
-
+  public refreshLayout(layoutParams: LayerLayoutParams) {
     // Check the heights of each row, in case different layers have different row counts.
+    const layerHeight = layoutParams.keyboardHeight;
     const nRows = this.rows.length;
     const rowHeight = this._rowHeight = Math.floor(layerHeight/(nRows == 0 ? 1 : nRows));
 
-    if(vkbd.usesFixedHeightScaling) {
+    const usesFixedWidthScaling = layoutParams.widthStyle.absolute;
+    if(usesFixedWidthScaling) {
       this.element.style.height=(layerHeight)+'px';
     }
 
+    const spacebarTextClosure = this.showLanguage(layoutParams.spacebarText);
+
+    // Update row layout properties
+    const rowClosures: (() => void)[] = [];
     for(let nRow=0; nRow<nRows; nRow++) {
       const oskRow = this.rows[nRow];
       const bottom = (nRows-nRow-1)*rowHeight+1;
 
-      if(vkbd.usesFixedHeightScaling) {
+      if(usesFixedWidthScaling) {
         // Calculate the exact vertical coordinate of the row's center.
         this.spec.row[nRow].proportionalY = ((layerHeight - bottom) - rowHeight/2) / layerHeight;
+      }
 
-        if(nRow == nRows-1) {
+      let rowClosure = oskRow.refreshLayout(layoutParams);
+      if(nRow == nRows-1) {
+        const oldRowClosure = rowClosure;
+        rowClosure = () => {
+          oldRowClosure();
           oskRow.element.style.bottom = '1px';
-        }
+        };
       }
+      rowClosures.push(rowClosure);
+    }
 
-      oskRow.refreshLayout(vkbd);
+    const rowKeyClosures: (() => void)[] = [];
+    for(const row of this.rows) {
+      const batchedUpdates = row.refreshKeyLayouts(layoutParams);
+      rowKeyClosures.push(batchedUpdates);
     }
+
+    // After row layout properties have been updated, _then_ update key internals.
+    // Doing this separately like this helps to reduce layout reflow.
+    spacebarTextClosure();
+    rowClosures.forEach((closure) => closure());
+    rowKeyClosures.forEach((closure) => closure());
```

Comment:
I'm really not keen on all this passing around of closures. It's really hard to follow what is happening when. It shouldn't be necessary with a good structure to the code.

Reply:
Huh.

To me, I felt that this made it extremely clear that we've deferred and essentially batched all operations that could affect DOM layout (and thus, could trigger layout reflow) to this one place.  (Granted, maybe I should have added documentation of this detail to the lower-level closure-building calls...)  All of the closures rely on precomputed values, some of which do refer to existing layout (when prerequisite values are uncached).  By precomputing all values first, without any edits... and then triggering those edits later... that's how we prevent layout reflow here.

Now that all precomputations are complete, we're safe to apply all of them sequentially without any reference to layout or style information.  Use of closures automatically caches the precomputations with manually-defined code, which is a major reason I used the pattern.  (Rather than manually tracking the values, we can just leverage existing language features to do that heavy lifting.)

---

Granted, maybe some extra comments could make this clearer - whether here or on the lower-level `refresh` methods in regard to why the returned closures exist.  What would make this clearer for you?  Alternatively... what alternate "good structure" would you suggest?